File: taxonomist/twitter.py
from datetime import datetime
from time import sleep
import logging

from requests_oauthlib import OAuth1Session
import requests


class Client(object):
    """Adapter for hitting the Twitter API."""
    BASE_URL = 'https://api.twitter.com'
    USERS_LOOKUP_CHUNK_SIZE = 100

    def __init__(self, api_key, api_secret):
        self.api_key = api_key
        self.api_secret = api_secret

    def request_token(self):
        oauth = OAuth1Session(self.api_key,
                              self.api_secret,
                              callback_uri='http://127.0.0.1:5000/callback')

        return oauth.fetch_request_token(self.url_for('/oauth/request_token'))

    def access_token(self, oauth_token, oauth_token_secret, oauth_verifier):
        oauth = OAuth1Session(self.api_key,
                              self.api_secret,
                              resource_owner_key=oauth_token,
                              resource_owner_secret=oauth_token_secret,
                              verifier=oauth_verifier)

        return oauth.fetch_access_token(self.url_for('/oauth/access_token'))

    def url_for(self, endpoint):
        return self.BASE_URL + endpoint


class AuthedClient(Client):
    def __init__(self, api_key, api_secret, oauth_token, oauth_token_secret):
        super(AuthedClient, self).__init__(api_key, api_secret)

        self.oauth = OAuth1Session(api_key,
                                   api_secret,
                                   resource_owner_key=oauth_token,
                                   resource_owner_secret=oauth_token_secret)

    def direct_messages_sent(self, since_id=None, max_id=None):
        params = {'since_id': since_id,
                  'max_id': max_id,
                  'count': 200,
                  'include_entities': 'false'}
        response = self.http(self.oauth.get,
                             '/1.1/direct_messages/sent.json',
                             params=params)
        return response.json()

    def favorites_list(self, user_id, since_id=None, max_id=None):
        params = {'user_id': user_id,
                  'since_id': since_id,
                  'max_id': max_id,
                  'count': 200,
                  'include_entities': 'false'}
        response = self.http(self.oauth.get,
                             '/1.1/favorites/list.json',
                             params=params)
        return response.json()

    def friends_ids(self, user_id):
        """Retrieve the first 5000 friend IDs for the given user."""
        params = {'user_id': user_id}
        response = self.http(self.oauth.get,
                             '/1.1/friends/ids.json',
                             params=params)
        return response.json().get('ids')

    def lists_ownerships(self, user_id, cursor=None):
        params = {'user_id': user_id, 'count': 1000, 'cursor': cursor}
        response = self.http(self.oauth.get,
                             '/1.1/lists/ownerships.json',
                             params=params)
        return response.json()

    def users_lookup(self, user_ids):
        params = {'user_id': ','.join([str(id) for id in user_ids])}
        response = self.http(self.oauth.post,
                             '/1.1/users/lookup.json',
                             data=params)
        return response.json()

    def statuses_user_timeline(self, user_id, since_id=None, max_id=None):
        # Sadly, there's no good way of checking the retweet chain, so
        # we can just ignore RTs for now.
        params = {'user_id': user_id,
                  'since_id': since_id,
                  'max_id': max_id,
                  'count': 200,
                  'trim_user': 'true',
                  'include_rts': 'false'}
        response = self.http(self.oauth.get,
                             '/1.1/statuses/user_timeline.json',
                             params=params)
        return response.json()

    def http(self, func, endpoint, **kwargs):
        response = func(self.url_for(endpoint), **kwargs)
        if response.status_code == requests.codes.too_many_requests:
            raise RateLimitError(response)
        response.raise_for_status()
        return response


def retry_rate_limited(f):
    def retry(*args, **kwargs):
        logger = logging.getLogger('taxonomist')
        while True:
            result = None
            try:
                result = f(*args, **kwargs)
            except RateLimitError as exc:
                delta = exc.rate_limit_reset - datetime.now()
                countdown = delta.total_seconds() + 1
                fmt = "Rate limited trying to hit %s, sleeping for %s seconds"
                logger.warn(fmt, exc.response.url, countdown)
                sleep(countdown)
            return result
    return retry


class RateLimitError(Exception):
    def __init__(self, response):
        self.response = response

    def __str__(self):
        return repr(self.response)

    @property
    def rate_limit_reset(self):
        reset = self.response.headers['x-rate-limit-reset']
        return datetime.fromtimestamp(int(reset))

File: lib/taxonomist/twitter.rb
require "faraday"
require "faraday_middleware"

module Taxonomist
  class Twitter
    class RateLimitedError < StandardError
      attr_reader :reset_at

      def initialize(reset_at:)
        @reset_at = reset_at
      end
    end

    attr_reader *%i[ api_key api_secret conn ]

    def initialize(api_key:, api_secret:)
      @api_key, @api_secret = api_key, api_secret
    end

    class OAuth < Twitter
      def request_token(callback:)
        conn = Faraday.new("https://api.twitter.com/oauth") do |conn|
          conn.request :oauth, consumer_key: api_key,
                               consumer_secret: api_secret,
                               callback: callback

          conn.response :raise_error

          conn.adapter Faraday.default_adapter
        end

        resp = conn.post("request_token")
        Faraday::Utils.parse_query(resp.body)
      end

      def access_token(token:, token_secret:, oauth_verifier:)
        conn = Faraday.new("https://api.twitter.com/oauth") do |conn|
          conn.request :url_encoded
          conn.request :oauth, consumer_key: api_key,
                               consumer_secret: api_secret,
                               token: token,
                               token_secret: token_secret

          conn.response :raise_error

          conn.adapter Faraday.default_adapter
        end

        resp = conn.post("access_token", oauth_verifier: oauth_verifier)
        Faraday::Utils.parse_query(resp.body)
      end
    end

    class Authed < Twitter
      attr_reader *%i[ access_token access_token_secret ]

      def initialize(api_key:, api_secret:,
                     access_token:, access_token_secret:)
        super(api_key: api_key, api_secret: api_secret)

        @access_token, @access_token_secret = access_token, access_token_secret

        @conn = Faraday.new("https://api.twitter.com/1.1") do |conn|
          conn.request :oauth, consumer_key: api_key,
                               consumer_secret: api_secret,
                               token: access_token,
                               token_secret: access_token_secret
          conn.request :json

          conn.response :raise_error
          conn.response :json, :content_type => /\bjson$/

          conn.adapter Faraday.default_adapter
        end
      end

      def friends_ids(user_id:)
        resp = get("friends/ids.json", user_id: user_id)
        cursored(resp.body["ids"], resp)
      end

      def lists_members(list_id:)
        resp = get("lists/members.json", list_id: list_id, count: 5_000)
        cursored(resp.body["users"], resp)
      end

      def lists_ownerships(user_id:)
        resp = get("lists/ownerships.json", user_id: user_id, count: 1_000)
        cursored(resp.body["lists"], resp)
      end

      def statuses_user_timeline(user_id:, since_id: nil, max_id: nil)
        params = { user_id: user_id, count: 200, trim_user: true }
        params[:since_id] = since_id if since_id
        params[:max_id] = max_id if max_id
        resp = get('statuses/user_timeline.json', params)
        resp.body
      end

      def users_lookup(user_ids:)
        user_ids = user_ids.join(?,)
        get("users/lookup.json", user_id: user_ids).body
      end

      def users_show(user_id:)
        get("users/show.json", user_id: user_id).body
      end

      private

      def get(endpoint, **kwargs)
        self.conn.get(endpoint, **kwargs)
      rescue Faraday::ClientError => e
        response = e.response
        if response[:status] == 429
          reset_at = Time.at(response[:headers]['x-rate-limit-reset'].to_i)
          raise RateLimitedError.new(reset_at: reset_at)
        end
        raise
      end

      def cursored(obj, resp)
        obj.extend(Cursored)
        obj.next_cursor = resp.body["next_cursor"]
        obj.previous_cursor = resp.body["previous_cursor"]
        obj
      end

      module Cursored
        attr_accessor *%i[ next_cursor previous_cursor ]
      end
    end
  end
end
